docs/TypeInference.md

```diff
@@ -0,0 +1,218 @@
+    These records have fields `_time`, `_start`, `_stop`, and `_value`.
+    The `_value` field can be of any type.
+    All other fields represented by the row variable `'s` must be of type _string_.
+    Note the last part corresponds to the tag values of a series.
+
+    An example of a program that will fail to compile is:
+    ```
+    from(bucket: "telegraf")
+        |> range(start: -1h)
+        |> filter(fn: (r) => r.host == 1) // error: {host: int | 't} is not of kind StrRec
+    ```
+    The reason for failed compilation will be clear soon, but basically `range` and `filter` are just pass through transformations.
+    They don't extend or modify the types of their inputs.
+    As a result `filter` expects to receive the same list type that `from` returns and so it expects `r.host` to be of type _string_.
+    This contradicts the usage of `r.host` as an _int_.
+
+* range
+
+    ```
+    forall ['r] where 'r:Rec (start: time, stop: time, tables: [ {_start: time | _stop: time | 'r} ]) -> [ {_start: time | _stop: time | 'r} ]
+    ```
+    Let's focus on the `tables` parameter.
+    `range` takes a `tables` parameter that is a list of records that must have fields `_start` and `_stop` and returns the same list type.
+
+* filter
+
+    ```
+    forall ['r] where 'r:Rec (fn: (r: 'r) -> bool, tables: ['r]) -> ['r]
+    ```
+    `filter` takes a list of records and returns that same list of records.
+
+    From the type signatures of `range` and `filter` it should be clear why the above example program fails unification.
+
+* map
+
+    ```
+    forall ['r, 's] where 'r:Rec, 's:Rec (fn: ('r) => 's, tables: [ 'r ]) -> [ 's ]
+    ```
+
+* keep
+
+    `keep` and `drop` operations can be implemented using `map`:
+    ```
+    map(fn: (r) => ( {a: r.a | b: r.b | {} } )) // keep columns a and b
+    map(fn: (r) => ( {r - a, b} ))              // drop columns a and b
+    ```
+    Assigning a type signature to the actual `keep` and `drop` functions unfortunatly is not so simple.
+    These functions take the names of the columns they operate on via a `columns` parameter.
+    Assigning an accurate type signature to these functions requires the notion of compile time constants or label types [#1122](https://github.com/influxdata/flux/issues/1122).
+
+    Another option is a preprocessing step before type inference that rewrites the Flux AST by replacing calls to `keep` and `drop` with their `map` equivalents.
+
+* aggregates
+
+    Assigning accurate types to aggregates is also a bit challenging.
+    This is due to the fact the aggregates will implicitly drop columns that are not part of the group key.
+    However if aggregates didn't drop columns dynamically, we could type them like so:
+    ```
+    forall ['r] (column: str, tables: [ {_time: time | 'r} ]) -> [ 'r ]
+    ```
+    In words, aggregates take a list of records with a `_time` fields and return that same list minus the `_time` field.
```

Comment:
Should we look into what it would take to model the group key in the type system? Maybe just a kinding constraint on the fields? Even if we did how would we model the dynamic behavior of dropping all non group keys fields?

Also aggregates always produce a `_value` column. So its really:

```
 forall ['r] (column: str, tables: [ {_time: time | 'r} ]) -> [ {_value: TYPE | 'r} ]
```

Where `TYPE` depends on the specific aggregate function.

Reply:
As I mentioned above, instead of conflating record extension and update, perhaps it would be best for aggregates to take in a record with `_time` and `_value` fields and return a list of the same record type?

As for modeling the group key, I'm not totally sure yet. A kind constraint might work. Another option is instead of treating a stream like a list of records, treat is as a record with a `group_key` property (that is again another record) along with a `data` property which is a list of records. This I think would force us into schema homogeneity though as you'd somehow have to specify what the group key from `from` should be. So still a lot of unknowns here. At least for me anyway.